### PR TITLE
[Fix]: Enable support for Arm Neon & SVE support for FP32 Gemm Wrapper

### PR DESCRIPTION
**Pull request by nikhil-arm**
**Performance Improvements**:
Linear Layer [ 1x512 * 512x512 ] ->  2x - 4x
Linear Layer [ 3x512 * 512x512 ] -> 2x - 4x

cc @malfet @snadampal @milpuz01 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @BoyuanFeng

### CODE REVIEW

**Review by leslie-fang-intel — APPROVED** [review]
Thanks for the PR. Overall LGTM. Not sure if current CI for CPP GEMM Template will run on with `VecNEON` and `VecSVE`, if not we may need to enable it to cover the testing.

**Review by leslie-fang-intel — COMMENTED** [review]
Let's check if CI covers ARM platform. Thanks.

**Review by jgong5 — APPROVED** [review]

**Comment by ng-05** [comment]
@pytrorchbot label "ciflow/linux-aarch64"

**Comment by ng-05** [comment]
@leslie-fang-intel can you please approve these changes

**Review by cfRod — APPROVED** [review]

**Review by leslie-fang-intel — APPROVED** [review]
Thanks for the PR, LGTM. Please check all the CPU FP32 GEMM template related UTs pass on Arm.

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
@ng-05 can you please share the script that you've run to observe those perf improvements?

**Comment by ng-05** [comment]
> @ng-05 can you please share the script that you've run to observe those perf improvements?

Run Command: 
```
LD_PRELOAD=/usr/lib/aarch64-linux-gnu/libtcmalloc.so.4 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKEND=CPP  TORCHINDUCTOR_MAX_AUTOTUNE=1 TORCHINDUCTOR_CPP_WRAPPER=1  TORCHINDUCTOR_FREEZING=1  OMP_NUM_THREADS=16 python test_inductor_linear.py
```
```
 import torch
import torch.nn as nn
import torch.profiler as profiler
import time
# Enable torch.no_grad globally
torch.set_grad_enabled(False)

# Define models as nn.Modules
class LinearModel(nn.Module):
    def __init__(self, input_size, output_size, bias):
        super(LinearModel, self).__init__()
        self.linear = nn.Linear(input_size, output_size, bias=bias)

    def forward(self, x):
        return self.linear(x)

# Create and compile models
linear_3x512_model = torch.compile(LinearModel(512, 512, bias=False), mode="max-autotune", fullgraph=True)
linear_1x512_model = torch.compile(LinearModel(512, 512, bias=True), mode="max-autotune", fullgraph=True)

# Generate random inputs
input_3x512 = torch.randn(3, 512)
input_1x512 = torch.randn(1, 512)

# Number of iterations for benchmarking
num_iterations = 1000

# Warm-up function
def warmup(model, input_tensor):
    for _ in range(100):  # Warm-up phase to stabilize performance
        _ = model(input_tensor)

# Benchmark function
def benchmark(model, input_tensor):
    start_time = time.time()

    for _ in range(num_iterations):
        _ = model(input_tensor)

    end_time = time.time()
    return (end_time - start_time) / num_iterations

# Profile function
def profile_layer(model, input_tensor):
    with profiler.profile(
        activities=[
            profiler.ProfilerActivity.CPU],
        on_trace_ready=profiler.tensorboard_trace_handler('./log'),
        record_shapes=True,
        profile_memory=True
    ) as prof:
        for _ in range(10):  # Reduced iterations for profiling
            _ = model(input_tensor)

    print(prof.key_averages().table(sort_by="cpu_time_total", row_limit=10))

# Warm-up
print("Warming up...")
warmup(linear_3x512_model, input_3x512)
warmup(linear_1x512_model, input_1x512)

# Benchmark without profiler
print("Benchmarking...")
average_time_3x512 = benchmark(linear_3x512_model, input_3x512)
average_time_1x512 = benchmark(linear_1x512_model, input_1x512)

print(f"Average execution time for Linear Layer [3x512 * 512x512]: {average_time_3x512 * 1e6:.2f} microseconds")
print(f"Average execution time for Linear Layer [1x512 * 512x512]: {average_time_1x512 * 1e6:.2f} microseconds")

# Profile with torch.profiler
print("Profiling Linear Layer [3x512 * 512x512]...")
profile_layer(linear_3x512_model, input_3x512)

print("Profiling Linear Layer [1x512 * 512x512]...")
profile_layer(linear_1x512_model, input_1x512)

```

**Comment by ng-05** [comment]
I have updated the speedup numbers in PR description. I am not able to get the 10x improvement with single linear layers but with a genAI model I can see the 10x improvements for certain shapes.
